Add `BufRead::skip_until`

WilliamVenner opened the pull request:
Alternative version of `BufRead::read_until` that simply discards data, rather than copying it into a buffer.

Useful for situations like skipping irrelevant data in a binary file format that is NUL-terminated.

<details>
<summary>Benchmark</summary>

```
running 2 tests
test bench_read_until ... bench:         123 ns/iter (+/- 6)
test bench_skip_until ... bench:          66 ns/iter (+/- 3)
```

```rs
#![feature(test)]
extern crate test;
use test::Bencher;

use std::io::{ErrorKind, BufRead};

fn skip_until<R: BufRead + ?Sized>(r: &mut R, delim: u8) -> Result<usize, std::io::Error> {
    let mut read = 0;
    loop {
        let (done, used) = {
            let available = match r.fill_buf() {
                Ok(n) => n,
                Err(ref e) if e.kind() == ErrorKind::Interrupted => continue,
                Err(e) => return Err(e),
            };
            match memchr::memchr(delim, available) {
                Some(i) => (true, i + 1),
                None => (false, available.len()),
            }
        };
        r.consume(used);
        read += used;
        if done || used == 0 {
            return Ok(read);
        }
    }
}

const STR: &[u8] = b"Ferris\0Hello, world!\0";

#[bench]
fn bench_skip_until(b: &mut Bencher) {
    b.iter(|| {
        let mut io = std::io::Cursor::new(test::black_box(STR));
        skip_until(&mut io, b'\0').unwrap();
        let mut hello = Vec::with_capacity(b"Hello, world!\0".len());
        let num_bytes = io.read_until(b'\0', &mut hello).unwrap();
        assert_eq!(num_bytes, b"Hello, world!\0".len());
        assert_eq!(hello, b"Hello, world!\0");
    });
}


#[bench]
fn bench_read_until(b: &mut Bencher) {
    b.iter(|| {
        let mut io = std::io::Cursor::new(test::black_box(STR));
        io.read_until(b'\0', &mut Vec::new()).unwrap();
        let mut hello = Vec::with_capacity(b"Hello, world!\0".len());
        let num_bytes = io.read_until(b'\0', &mut hello).unwrap();
        assert_eq!(num_bytes, b"Hello, world!\0".len());
        assert_eq!(hello, b"Hello, world!\0");
    });
}
```
</details>

WilliamVenner committed:
Add `BufRead::skip_until`

rust-highfive commented:
Thanks for the pull request, and welcome! The Rust team is excited to review your changes, and you should hear from @thomcc (or someone else) soon.

Please see [the contribution instructions](https://rustc-dev-guide.rust-lang.org/contributing.html) for more information.

WilliamVenner commented:
@rustbot label +T-libs-api -T-libs

thomcc commented:
Hi, thanks for the PR.

As the bot mentioned, you should file an API change proposal for this. This should be done here: https://github.com/rust-lang/libs-team/issues/new?assignees=&labels=api-change-proposal%2C+T-libs-api&template=api-change-proposal.md&title=%28My+API+Change+Proposal%29

See also: <https://std-dev-guide.rust-lang.org/feature-lifecycle/api-change-proposals.html>, which describes the process.

pitaj commented:
ACP here: https://github.com/rust-lang/libs-team/issues/63

Do create a tracking issue though.

thomcc commented:
Can you add some tests for this?

@rustbot author

WilliamVenner committed:
Add `BufRead::skip_until` test

WilliamVenner committed:
Move `BufRead::skip_until` test to a more appropriate location

WilliamVenner commented:
OK - added a test. There's also the doc comment test which is already present.

Is there any reason for the `#[cfg_attr(target_os = "emscripten", ignore)]` on the adjacent `read_until` test? Will that need to be added to the `skip_until` test too? 🤔

WilliamVenner committed:
Add tracking issue ID for `bufread_skip_until` feature gate

thomcc commented on library/std/src/io/tests.rs:
```diff
@@ -25,6 +25,34 @@ fn read_until() {
+        }
+
+        // skip past `ignore\0`
+        reader.skip_until(0).unwrap();
```
Can you assert that `skip_until` is returning the right length?

thomcc reviewed (dismissed):
One nit. Otherwise looks fine.

thomcc commented:
@rustbot author

WilliamVenner committed:
Add assertion to test `skip_until` return value

The extra `\0` in this commit is needed because the assertion on line 49 will fail otherwise (as `skip_until` stops reading on EOF and therefore does not read a trailing `\0`, returning 6 read bytes rather than the expected 7)

WilliamVenner commented:
@rustbot ready

dtolnay reviewed (approved):
Thank you!

dtolnay commented:
@bors r+

bors commented:
:pushpin: Commit 7c1ab71f71087d88aade79925ae68a447397422f has been approved by `dtolnay`

It is now in the [queue](https://bors.rust-lang.org/homu/queue/rust) for this repository.

<!-- @bors r=dtolnay 7c1ab71f71087d88aade79925ae68a447397422f -->
<!-- homu: {"type":"Approved","sha":"7c1ab71f71087d88aade79925ae68a447397422f","approver":"dtolnay","queue":"https://bors.rust-lang.org/homu/queue/rust"} -->

bors commented:
:hourglass: Testing commit 7c1ab71f71087d88aade79925ae68a447397422f with merge e68f935117fef05e7663605a5a6671a6bb4ce719...
<!-- homu: {"type":"BuildStarted","head_sha":"7c1ab71f71087d88aade79925ae68a447397422f","merge_sha":"e68f935117fef05e7663605a5a6671a6bb4ce719"} -->

bors commented:
:sunny: Test successful - [checks-actions](https://github.com/rust-lang-ci/rust/actions/runs/6974746253/job/18982420223)
Approved by: dtolnay
Pushing e68f935117fef05e7663605a5a6671a6bb4ce719 to master...
<!-- homu: {"type":"BuildCompleted","approved_by":"dtolnay","base_ref":"master","builders":{"checks-actions":"https://github.com/rust-lang-ci/rust/actions/runs/6974746253/job/18982420223"},"merge_sha":"e68f935117fef05e7663605a5a6671a6bb4ce719"} -->

rust-timer commented:
Finished benchmarking commit (e68f935117fef05e7663605a5a6671a6bb4ce719): [comparison URL](https://perf.rust-lang.org/compare.html?start=a1a37735cbc3db359d0b24ba9085c9fcbe1bc274&end=e68f935117fef05e7663605a5a6671a6bb4ce719&stat=instructions:u).

### Overall result: no relevant changes - no action needed


@rustbot label: -perf-regression



### [Instruction count](https://perf.rust-lang.org/compare.html?start=a1a37735cbc3db359d0b24ba9085c9fcbe1bc274&end=e68f935117fef05e7663605a5a6671a6bb4ce719&stat=instructions:u)
This benchmark run did not return any relevant results for this metric.

### [Max RSS (memory usage)](https://perf.rust-lang.org/compare.html?start=a1a37735cbc3db359d0b24ba9085c9fcbe1bc274&end=e68f935117fef05e7663605a5a6671a6bb4ce719&stat=max-rss)
<details>
<summary>Results</summary>

This is a less reliable metric that may be of interest but was not used to determine the overall result at the top of this comment.

|                                    | mean | range        | count |
|:----------------------------------:|:----:|:------------:|:-----:|
| Regressions ❌ <br /> (primary)    | -    | -            | 0     |
| Regressions ❌ <br /> (secondary)  | 2.9% | [2.9%, 2.9%] | 1     |
| Improvements ✅ <br /> (primary)   | -    | -            | 0     |
| Improvements ✅ <br /> (secondary) | -    | -            | 0     |
| All ❌✅ (primary)                 | -    | -            | 0     |
</details>

### [Cycles](https://perf.rust-lang.org/compare.html?start=a1a37735cbc3db359d0b24ba9085c9fcbe1bc274&end=e68f935117fef05e7663605a5a6671a6bb4ce719&stat=cycles:u)
<details>
<summary>Results</summary>

This is a less reliable metric that may be of interest but was not used to determine the overall result at the top of this comment.

|                                    | mean  | range          | count |
|:----------------------------------:|:-----:|:--------------:|:-----:|
| Regressions ❌ <br /> (primary)    | -     | -              | 0     |
| Regressions ❌ <br /> (secondary)  | -     | -              | 0     |
| Improvements ✅ <br /> (primary)   | -     | -              | 0     |
| Improvements ✅ <br /> (secondary) | -6.8% | [-6.8%, -6.8%] | 1     |
| All ❌✅ (primary)                 | -     | -              | 0     |
</details>

### [Binary size](https://perf.rust-lang.org/compare.html?start=a1a37735cbc3db359d0b24ba9085c9fcbe1bc274&end=e68f935117fef05e7663605a5a6671a6bb4ce719&stat=size:linked_artifact)
This benchmark run did not return any relevant results for this metric.

**Bootstrap**: 675.978s -> 675.904s (-0.01%)
**Artifact size**: 313.73 MiB -> 313.71 MiB (-0.01%)
